cmd/kubeadm/app/util/config/common_test.go

```diff
@@ -439,6 +447,94 @@ func TestMigrateOldConfig(t *testing.T) {
 	}
 }
 
+// NOTE: do not delete this test once an older API is removed and there is only one API left.
+// Update the inline "gv" and "gvExperimental" variables, to have the GroupVersion String of
+// the API to be tested. If there are no experimental APIs make "gvExperimental" point to
+// an non-experimental API.
+func TestValidateConfig(t *testing.T) {
+	var (
+		gv             = kubeadmapiv1old.SchemeGroupVersion.String()
+		gvExperimental = kubeadmapiv1.SchemeGroupVersion.String()
+	)
+	tests := []struct {
+		name              string
+		cfg               string
+		expectedError     bool
+		allowExperimental bool
+	}{
+		{
+			name: "invalid subdomain",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			nodeRegistration:
+			  criSocket: %s
+			  name: foo bar # not a valid subdomain
+			`, gv, constants.UnknownCRISocket)),
+			expectedError: true,
+		},
+		{
+			name: "unknown API GVK",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: foo/bar # not a valid GroupVersion
+			kind: zzz # not a valid Kind
+			nodeRegistration:
+			  criSocket: %s
+			`, constants.UnknownCRISocket)),
+			expectedError: true,
+		},
+		{
+			name: "legacy API GVK",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: kubeadm.k8s.io/v1beta1 # legacy API
+			kind: InitConfiguration
+			nodeRegistration:
+			  criSocket: %s
+			`, constants.UnknownCRISocket)),
+			expectedError: true,
+		},
+		{
+			name: "unknown field",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			foo: bar
+			nodeRegistration:
+			  criSocket: %s
+			`, gv, constants.UnknownCRISocket)),
+			expectedError: true,
+		},
+		{
+			name: "valid",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			nodeRegistration:
+			  criSocket: %s
+			`, gv, constants.UnknownCRISocket)),
+			expectedError: false,
+		},
+		{
+			name: "valid: experimental API",
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			`, gvExperimental)),
```

Comment:
A test case that is invalid due to `Experimental` not allowed, may be needed.

Reply:
good catch, added experimental=false, expectedError=true test cases for both validate and migrate.

Reply:
LGTM